notebooks/utils/estimation.py

```diff
@@ -0,0 +1,229 @@
+        elif filter_type == 'highpass':
+            b = firwin(numtaps=order + 1, cutoff=frequencies, fs=fs, pass_zero=False, **kwargs)
+        elif filter_type == 'lowpass':
+            b = firwin(numtaps=order + 1, cutoff=frequencies, fs=fs, pass_zero=True, **kwargs)
+        a = 1
+        filter_args = (b, a)
+    else:  # filter_design == 'iir':
+        kwargs.setdefault('output', 'ba')
+        filter_args = iirfilter(order, Wn=frequencies, btype=filter_type, fs=fs, **kwargs)
+    return filter_args
+
+
+class SSVEPCovariance(BaseEstimator, TransformerMixin):
+    """Estimate special form covariance matrix for SSVEP.
+
+    Perform a simple covariance matrix estimation for each given trial.
+
+    Parameters
+    ----------
+    estimator : string (default: 'scm')
+        covariance matrix estimator. For regularization consider 'lwf' or 'oas'
+        For a complete list of estimator, see `utils.covariance`.
+
+    """
+
+    def __init__(self, rate: float = 256, flickering_frequencies: tuple = (13, 17, 21),
+                 estimator: str = 'scm', **kwargs):
+        """Init."""
+        self._estimator = estimator
+        self._design_filters(rate, flickering_frequencies, **kwargs)
+
+    def fit(self, X: np.array, y: np.array = None) -> object:
+        """Fit.
+
+        Do nothing. For compatibility purpose.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_trials, n_channels, n_samples)
+            ndarray of trials.
+        y : ndarray shape (n_trials,)
+            labels corresponding to each trial, not used.
+
+        Returns
+        -------
+        self : SSVEPCovariance instance
+            The SSVEPCovariance instance.
+        """
+        return self
+
+    def transform(self, X: np.array) -> np.array:
+        """Estimate covariance matrices.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_trials, n_channels, n_samples)
+            ndarray of trials.
+
+        Returns
+        -------
+        covmats : ndarray, shape (n_trials, n_channels * n_frequencies, n_channels * n_frequencies)
+            SSVEP covariance matrices for each trials.
+        """
+        filtered = []
+        for filter in self.filters:
+            filtered_X = filter(X, axis=-1)
+            filtered.append(filtered_X)
+        mega_X = np.concatenate(filtered, axis=1)
+        out = covariances(mega_X, estimator=self._estimator)
+        return out
+
+    def _design_filters(self, fs: float, flickering_frequencies: tuple,
+                        bandwidth: float = 4, order: int = 2,
+                        filter_design: str = 'iir', padlen: int = None):
+        self.filters = []
+        for flickering_frequency in flickering_frequencies:
+            frequencies = [flickering_frequency - bandwidth / 2,
+                           flickering_frequency + bandwidth / 2]  # TODO/Question : shall we consider harmonics as well ?
```

Comment:
Target frequencies are sufficient. But in theory, adding harmonics enhances the signal-to-noise ratio of SSVEP based BCIs. It should be great to measure if this improvement is significant.